File: python/semantic_kernel/processes/process_types.py
# Copyright (c) Microsoft. All rights reserved.

import types
from typing import Any, TypeVar, Union, get_args, get_origin, get_type_hints

from semantic_kernel.processes.kernel_process.kernel_process_step import KernelProcessStep
from semantic_kernel.utils.feature_stage_decorator import experimental

TStep = TypeVar("TStep", bound=KernelProcessStep)
TState = TypeVar("TState")


@experimental
def get_generic_state_type(cls) -> Any:
    """Given a subclass of KernelProcessStep, retrieve the concrete type of 'state'."""
    try:
        type_hints = get_type_hints(cls)
        t_state = type_hints.get("state", None)
        if t_state is not None:
            # Handle Union types and Python 3.12 '|' syntax explicitly
            origin = get_origin(t_state)
            if origin in {Union, types.UnionType}:
                args = get_args(t_state)
                # Filter out 'NoneType' and keep the remaining types
                non_none_args = [arg for arg in args if arg is not type(None)]
                t_state = non_none_args[0] if non_none_args else None
            # Check if t_state is a TypeVar (e.g., TypeVar('TState'))
            if isinstance(t_state, TypeVar):
                t_state = None  # We don't have a concrete type
            return t_state
    except Exception:
        pass  # nosec
    # Recursively check base classes
    for base in cls.__bases__:
        t_state = get_generic_state_type(base)
        if t_state is not None:
            return t_state
    return None

File: nodejs/semantic-kernel/processes/process-types.ts
import type { KernelProcessStep } from './kernel-process/kernel-process-step.js'

/**
 * Type variable for steps bound to KernelProcessStep.
 */
export type TStep = KernelProcessStep<any>

/**
 * Type variable for state.
 */
export type TState = any

/**
 * Given a subclass of KernelProcessStep, retrieve the concrete type of 'state'.
 * Note: TypeScript's runtime type information is limited, so this function
 * attempts to extract type information through reflection where possible.
 * @experimental
 * @param cls - The class to inspect
 * @returns The state type if it can be determined, otherwise null
 */
export function getGenericStateType(cls: new (...args: any[]) => any): any {
  try {
    // In TypeScript, we don't have runtime type information like Python's get_type_hints
    // This is a placeholder implementation. In practice, you might need to use
    // decorators or metadata to store type information at runtime, or rely on
    // explicit type parameters passed at construction time.

    // Check if the class has a static property that stores state type information
    // This would need to be set by a decorator or manually
    if ('__stateType__' in cls) {
      return (cls as any).__stateType__
    }

    // Recursively check base classes
    const proto = Object.getPrototypeOf(cls)
    if (proto && proto !== Function.prototype) {
      const baseType = getGenericStateType(proto)
      if (baseType !== null) {
        return baseType
      }
    }
  } catch {
    // Ignore errors
  }

  return null
}
